lib/Rx/Operator/ZipOperator.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+namespace Rx\Operator;
+
+use Rx\Disposable\CompositeDisposable;
+use Rx\ObservableInterface;
+use Rx\Observer\CallbackObserver;
+use Rx\ObserverInterface;
+use Rx\SchedulerInterface;
+
+class ZipOperator implements OperatorInterface {
+    /** @var ObservableInterface[] */
+    private $sources;
+    
+    /** @var callable */
+    private $resultSelector;
+
+    /** @var \SplQueue[] */
+    private $queues = [];
+
+    /** @var int */
+    private $completesRemaining = 0;
+
+    public function __construct(array $sources, callable $resultSelector = null)
+    {
+        $this->sources = $sources;
+
+        if ($resultSelector === null) {
+            $resultSelector = function () {
+                return func_get_args();
+            };
+        }
+        $this->resultSelector = $resultSelector;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function __invoke(
+        ObservableInterface $observable,
+        ObserverInterface $observer,
+        SchedulerInterface $scheduler = null
+    ) {
+        array_unshift($this->sources, $observable);
+
+        $disposable = new CompositeDisposable();
+
+        $this->completesRemaining = count($this->sources);
+
+        for ($i = 0; $i < count($this->sources); $i++) {
+            $this->queues[$i] = new \SplQueue();
+        }
+
+        for($i=0;$i<count($this->sources);$i++) {
+            $source = $this->sources[$i];
+
+            $disposable->add($source->subscribe(new CallbackObserver(
+                function ($x) use ($i, $observer) {
+                    // if there is another item in the sequence after one of the other source
+                    // observables completes, we need to complete at this time to match the
+                    // behavior of RxJS
+                    if ($this->completesRemaining < count($this->sources)) {
```

Comment:
Hm, not sure about this behavior. The sequence will now complete if:
- all sources are completed (makes sense)
- one source is completed and another source sends an `onNext`

The latter I find confusing. What is the idea behind it?